In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [3]:
master_dataframe = pd.read_csv('UFCMMABiggestDatasetWithDifferentials/masterdataframe.csv')
master_dataframe_b = master_dataframe[["date","fight_url", "event_url", "result", "fighter", "opponent", "division", "stance", "dob", "method", "total_comp_time", "round", "time",
                                        "fighter_url", "opponent_url", "referee", "time_format", "reach", "height", "age", "knockdowns", "sub_attempts", "reversals", "control", 
                                         "takedowns_landed", "takedowns_attempts", "sig_strikes_landed", "sig_strikes_attempts", "total_strikes_landed", "total_strikes_attempts", 
                                         "head_strikes_landed", "head_strikes_attempts", "body_strikes_landed", "body_strikes_attempts", "leg_strikes_landed", "leg_strikes_attempts",
                                         "distance_strikes_landed", "distance_strikes_attempts", "clinch_strikes_landed", "clinch_strikes_attempts", "ground_strikes_landed", 
                                         "ground_strikes_attempts"]]
display(master_dataframe_b.head())
print("Shape of Master Dataframe:", master_dataframe_b.shape)
master_dataframe_b.info()

date                                          fight_url  \
0  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   

                                           event_url  result         fighter  \
0  http://ufcstats.com/event-details/a6a9ab5a824e...       0  Sean Daugherty   
1  http://ufcstats.com/event-details/a6a9ab5a824e...       1    Scott Morris   
2  http://ufcstats.com/event-details/a6a9ab5a824e...       0      Ray Wizard   
3  http://ufcstats.com/event-details/a6a9ab5a824e...       1   Patrick Smith   
4  http://ufcstats.com/event-details/a6a9ab5a824e...       0   David Levicki   

         opponent     division    stance         dob  method  ...  \
0    Scott Morris  Open Weight       NaN  1975-12-04     SUB  ...   
1  Sean Daugherty  Open Weight  Orthodox         NaN     SUB  ...   
2   Patrick Smith  Open Weight       NaN         NaN     SUB  ...   
3      Ray Wizard  Open Weight  Orthodox  1963-08-28     SUB  ...   
4   Johnny Rhodes  Open Weight       NaN         NaN  KO/TKO  ...   

   body_strikes_landed  body_strikes_attempts leg_strikes_landed  \
0                    0                      0                  0   
1                    0                      0                  0   
2                    0                      0                  1   
3                    1                      1                  0   
4                    0                      0                  0   

  leg_strikes_attempts distance_strikes_landed distance_strikes_attempts  \
0                    2                       0                         3   
1                    0                       0                         0   
2                    1                       1                         1   
3                    0                       0                         0   
4                    0                       1                         2   

  clinch_strikes_landed  clinch_strikes_attempts  ground_strikes_landed  \
0                     0                        1                      0   
1                     1                        1                      0   
2                     0                        0                      0   
3                     1                        1                      0   
4                     2                        2                      1   

   ground_strikes_attempts  
0                        0  
1                        0  
2                        0  
3                        0  
4                        1  

[5 rows x 42 columns]

Shape of Master Dataframe: (13322, 42)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13322 entries, 0 to 13321
Data columns (total 42 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   date                       13322 non-null  object 
 1   fight_url                  13322 non-null  object 
 2   event_url                  13322 non-null  object 
 3   result                     13322 non-null  int64  
 4   fighter                    13322 non-null  object 
 5   opponent                   13322 non-null  object 
 6   division                   13322 non-null  object 
 7   stance                     13226 non-null  object 
 8   dob                        13087 non-null  object 
 9   method                     13322 non-null  object 
 10  total_comp_time            13322 non-null  int64  
 11  round                      13322 non-null  int64  
 12  time                       13322 non-null  object 
 13  fighter

In [4]:
print("\nMaster Dataframe Summary:")
display(master_dataframe_b.describe())


Master Dataframe Summary:


result  total_comp_time         round         reach  \
count  13322.000000     13322.000000  13322.000000  12039.000000   
mean       0.491142       628.426813      2.330431     72.136307   
std        0.499940       365.103370      1.013957      4.176705   
min        0.000000         5.000000      1.000000     58.000000   
25%        0.000000       272.000000      1.000000     70.000000   
50%        0.000000       794.000000      3.000000     72.000000   
75%        1.000000       900.000000      3.000000     75.000000   
max        1.000000      2160.000000      5.000000     84.000000   

             height           age    knockdowns  sub_attempts     reversals  \
count  13308.000000  13087.000000  13322.000000  13322.000000  13322.000000   
mean      70.383228     37.109345      0.215283      0.402792      0.136991   
std        3.482012      6.217236      0.493830      0.852478      0.423131   
min       60.000000     21.000000      0.000000      0.000000      0.000000   
25%       68.000000     33.000000      0.000000      0.000000      0.000000   
50%       71.000000     37.000000      0.000000      0.000000      0.000000   
75%       73.000000     41.000000      0.000000      1.000000      0.000000   
max       83.000000     79.000000      5.000000     10.000000      5.000000   

            control  ...  body_strikes_landed  body_strikes_attempts  \
count  13322.000000  ...         13322.000000           13322.000000   
mean     131.914953  ...             7.098709              10.299354   
std      174.784787  ...             8.225526              11.373791   
min        0.000000  ...             0.000000               0.000000   
25%        5.000000  ...             1.000000               2.000000   
50%       57.000000  ...             4.000000               7.000000   
75%      195.000000  ...            10.000000              15.000000   
max     1342.000000  ...           117.000000             133.000000   

       leg_strikes_landed  leg_strikes_attempts  distance_strikes_landed  \
count        13322.000000          13322.000000             13322.000000   
mean             5.664390              7.062903                24.714758   
std              7.590885              9.369506                27.564745   
min              0.000000              0.000000                 0.000000   
25%              1.000000              1.000000                 5.000000   
50%              3.000000              4.000000                15.000000   
75%              8.000000             10.000000                36.000000   
max             78.000000            101.000000               439.000000   

       distance_strikes_attempts  clinch_strikes_landed  \
count               13322.000000           13322.000000   
mean                   64.610344               5.095406   
std                    65.927645               7.381798   
min                     0.000000               0.000000   
25%                    14.000000               0.000000   
50%                    43.000000               2.000000   
75%                    96.000000               7.000000   
max                   737.000000              95.000000   

       clinch_strikes_attempts  ground_strikes_landed  ground_strikes_attempts  
count             13322.000000            13322.00000             13322.000000  
mean                  7.397238                5.04714                 7.392434  
std                  10.118685                9.00926                13.067610  
min                   0.000000                0.00000                 0.000000  
25%                   1.000000                0.00000                 0.000000  
50%                   4.000000                1.00000                 2.000000  
75%                  10.000000                6.00000                 9.000000  
max                 115.000000              100.00000               141.000000  

[8 rows x 28 columns]

In [5]:
print("\nMaster Dataframe:")
display(master_dataframe_b.head())


Master Dataframe:


date                                          fight_url  \
0  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   

                                           event_url  result         fighter  \
0  http://ufcstats.com/event-details/a6a9ab5a824e...       0  Sean Daugherty   
1  http://ufcstats.com/event-details/a6a9ab5a824e...       1    Scott Morris   
2  http://ufcstats.com/event-details/a6a9ab5a824e...       0      Ray Wizard   
3  http://ufcstats.com/event-details/a6a9ab5a824e...       1   Patrick Smith   
4  http://ufcstats.com/event-details/a6a9ab5a824e...       0   David Levicki   

         opponent     division    stance         dob  method  ...  \
0    Scott Morris  Open Weight       NaN  1975-12-04     SUB  ...   
1  Sean Daugherty  Open Weight  Orthodox         NaN     SUB  ...   
2   Patrick Smith  Open Weight       NaN         NaN     SUB  ...   
3      Ray Wizard  Open Weight  Orthodox  1963-08-28     SUB  ...   
4   Johnny Rhodes  Open Weight       NaN         NaN  KO/TKO  ...   

   body_strikes_landed  body_strikes_attempts leg_strikes_landed  \
0                    0                      0                  0   
1                    0                      0                  0   
2                    0                      0                  1   
3                    1                      1                  0   
4                    0                      0                  0   

  leg_strikes_attempts distance_strikes_landed distance_strikes_attempts  \
0                    2                       0                         3   
1                    0                       0                         0   
2                    1                       1                         1   
3                    0                       0                         0   
4                    0                       1                         2   

  clinch_strikes_landed  clinch_strikes_attempts  ground_strikes_landed  \
0                     0                        1                      0   
1                     1                        1                      0   
2                     0                        0                      0   
3                     1                        1                      0   
4                     2                        2                      1   

   ground_strikes_attempts  
0                        0  
1                        0  
2                        0  
3                        0  
4                        1  

[5 rows x 42 columns]

In [6]:
#MADE EDITS HERE
display(master_dataframe_b.head(20))
display(master_dataframe_b.columns)

date                                          fight_url  \
0   1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1   1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2   1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3   1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4   1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
5   1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
6   1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
7   1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
8   1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   
9   1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   
10  1994-03-11  http://ufcstats.com/fight-details/17ee4caf0698...   
11  1994-03-11  http://ufcstats.com/fight-details/17ee4caf0698...   
12  1994-03-11  http://ufcstats.com/fight-details/c53325be8ce3...   
13  1994-03-11  http://ufcstats.com/fight-details/c53325be8ce3...   
14  1994-03-11  http://ufcstats.com/fight-details/fcd8e8ec45eb...   
15  1994-03-11  http://ufcstats.com/fight-details/fcd8e8ec45eb...   
16  1994-03-11  http://ufcstats.com/fight-details/8ea72cf82fac...   
17  1994-03-11  http://ufcstats.com/fight-details/8ea72cf82fac...   
18  1994-03-11  http://ufcstats.com/fight-details/b72acedda7dd...   
19  1994-03-11  http://ufcstats.com/fight-details/b72acedda7dd...   

                                            event_url  result  \
0   http://ufcstats.com/event-details/a6a9ab5a824e...       0   
1   http://ufcstats.com/event-details/a6a9ab5a824e...       1   
2   http://ufcstats.com/event-details/a6a9ab5a824e...       0   
3   http://ufcstats.com/event-details/a6a9ab5a824e...       1   
4   http://ufcstats.com/event-details/a6a9ab5a824e...       0   
5   http://ufcstats.com/event-details/a6a9ab5a824e...       1   
6   http://ufcstats.com/event-details/a6a9ab5a824e...       0   
7   http://ufcstats.com/event-details/a6a9ab5a824e...       1   
8   http://ufcstats.com/event-details/a6a9ab5a824e...       0   
9   http://ufcstats.com/event-details/a6a9ab5a824e...       1   
10  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
11  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
12  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
13  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
14  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
15  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
16  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
17  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
18  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
19  http://ufcstats.com/event-details/a6a9ab5a824e...       1   

               fighter            opponent     division    stance         dob  \
0       Sean Daugherty        Scott Morris  Open Weight       NaN  1975-12-04   
1         Scott Morris      Sean Daugherty  Open Weight  Orthodox         NaN   
2           Ray Wizard       Patrick Smith  Open Weight       NaN         NaN   
3        Patrick Smith          Ray Wizard  Open Weight  Orthodox  1963-08-28   
4        David Levicki       Johnny Rhodes  Open Weight       NaN         NaN   
5        Johnny Rhodes       David Levicki  Open Weight  Orthodox         NaN   
6      Thaddeus Luster       Frank Hamaker  Open Weight       NaN         NaN   
7        Frank Hamaker     Thaddeus Luster  Open Weight       NaN         NaN   
8     Robert Lucarelli        Orlando Wiet  Open Weight       NaN         NaN   
9         Orlando Wiet    Robert Lucarelli  Open Weight  Southpaw         NaN   
10  Alberta Cerra Leon       Remco Pardoel  Open Weight       NaN         NaN   
11       Remco Pardoel  Alberta Cerra Leon  Open Weight  Southpaw         NaN   
12         Scott Baker       Jason DeLucia  Open Weight       NaN         NaN   
13       Jason DeLucia         Scott Baker  Open Weight  Sou

Index(['date', 'fight_url', 'event_url', 'result', 'fighter', 'opponent',
       'division', 'stance', 'dob', 'method', 'total_comp_time', 'round',
       'time', 'fighter_url', 'opponent_url', 'referee', 'time_format',
       'reach', 'height', 'age', 'knockdowns', 'sub_attempts', 'reversals',
       'control', 'takedowns_landed', 'takedowns_attempts',
       'sig_strikes_landed', 'sig_strikes_attempts', 'total_strikes_landed',
       'total_strikes_attempts', 'head_strikes_landed',
       'head_strikes_attempts', 'body_strikes_landed', 'body_strikes_attempts',
       'leg_strikes_landed', 'leg_strikes_attempts', 'distance_strikes_landed',
       'distance_strikes_attempts', 'clinch_strikes_landed',
       'clinch_strikes_attempts', 'ground_strikes_landed',
       'ground_strikes_attempts'],
      dtype='object')

### Missing Values

In [7]:
display(master_dataframe_b.isna().sum())
display(master_dataframe_b.count())

date                            0
fight_url                       0
event_url                       0
result                          0
fighter                         0
opponent                        0
division                        0
stance                         96
dob                           235
method                          0
total_comp_time                 0
round                           0
time                            0
fighter_url                     0
opponent_url                    0
referee                        64
time_format                     0
reach                        1283
height                         14
age                           235
knockdowns                      0
sub_attempts                    0
reversals                       0
control                         0
takedowns_landed                0
takedowns_attempts              0
sig_strikes_landed              0
sig_strikes_attempts            0
total_strikes_landed            0
total_strikes_

date                         13322
fight_url                    13322
event_url                    13322
result                       13322
fighter                      13322
opponent                     13322
division                     13322
stance                       13226
dob                          13087
method                       13322
total_comp_time              13322
round                        13322
time                         13322
fighter_url                  13322
opponent_url                 13322
referee                      13258
time_format                  13322
reach                        12039
height                       13308
age                          13087
knockdowns                   13322
sub_attempts                 13322
reversals                    13322
control                      13322
takedowns_landed             13322
takedowns_attempts           13322
sig_strikes_landed           13322
sig_strikes_attempts         13322
total_strikes_landed

## Out of 13322 items: 

# stance: missing 96 
As seen below, since the majority of values are within the Orthodox stance, the 96 missing values were replaced with Orthodox.

In [8]:
print(master_dataframe_b["stance"].value_counts()) #.fillna('1900-01-01', inplace=True)
master_dataframe_b["stance"].fillna('Orthodox', inplace=True)
display(master_dataframe_b.head(10))

stance
Orthodox       10007
Southpaw        2637
Switch           552
Open Stance       24
Sideways           6
Name: count, dtype: int64


/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/2020280991.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  master_dataframe_b["stance"].fillna('Orthodox', inplace=True)
/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/2020280991.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b["stance"

date                                          fight_url  \
0  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
5  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
6  1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
7  1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
8  1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   
9  1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   

                                           event_url  result  \
0  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
1  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
2  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
3  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
4  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
5  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
6  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
7  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
8  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
9  http://ufcstats.com/event-details/a6a9ab5a824e...       1   

            fighter          opponent     division    stance         dob  \
0    Sean Daugherty      Scott Morris  Open Weight  Orthodox  1975-12-04   
1      Scott Morris    Sean Daugherty  Open Weight  Orthodox         NaN   
2        Ray Wizard     Patrick Smith  Open Weight  Orthodox         NaN   
3     Patrick Smith        Ray Wizard  Open Weight  Orthodox  1963-08-28   
4     David Levicki     Johnny Rhodes  Open Weight  Orthodox         NaN   
5     Johnny Rhodes     David Levicki  Open Weight  Orthodox         NaN   
6   Thaddeus Luster     Frank Hamaker  Open Weight  Orthodox         NaN   
7     Frank Hamaker   Thaddeus Luster  Open Weight  Orthodox         NaN   
8  Robert Lucarelli      Orlando Wiet  Open Weight  Orthodox         NaN   
9      Orlando Wiet  Robert Lucarelli  Open Weight  Southpaw         NaN   

   method  ...  body_strikes_landed  body_strikes_attempts leg_strikes_landed  \
0     SUB  ...                    0                      0                  0   
1     SUB  ...                    0                      0                  0   
2     SUB  ...                    0                      0                  1   
3     SUB  ...                    1                      1                  0   
4  KO/TKO  ...                    0                      0                  0   
5  KO/TKO  ...                    1                      1                  1   
6     SUB  ...                    0                      0                  0   
7     SUB  ...                    0                      0                  0   
8  KO/TKO  ...                    0                      1                  1   
9  KO/TKO  ...                    1                      1                  0   

  leg_strikes_attempts distance_strikes_landed distance_strikes_attempts  \
0                    2                       0                         3   
1                    0                       0                         0   
2                    1                       1                         1   
3                    0                       0                         0   
4                    0                       1                         2   
5                    1                       1                         1   
6                    0                       0                         0   
7                    0                       1                         1   
8                    3                       2                         6   
9                    0                       1                    

# dob: missing 235

In [9]:
master_dataframe_b["dob"].fillna('1900-01-01', inplace=True)
display(master_dataframe_b.head(10))

/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/2272656395.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  master_dataframe_b["dob"].fillna('1900-01-01', inplace=True)
/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/2272656395.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b["dob"].fi

date                                          fight_url  \
0  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1  1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
2  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
3  1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
5  1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
6  1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
7  1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
8  1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   
9  1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   

                                           event_url  result  \
0  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
1  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
2  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
3  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
4  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
5  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
6  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
7  http://ufcstats.com/event-details/a6a9ab5a824e...       1   
8  http://ufcstats.com/event-details/a6a9ab5a824e...       0   
9  http://ufcstats.com/event-details/a6a9ab5a824e...       1   

            fighter          opponent     division    stance         dob  \
0    Sean Daugherty      Scott Morris  Open Weight  Orthodox  1975-12-04   
1      Scott Morris    Sean Daugherty  Open Weight  Orthodox  1900-01-01   
2        Ray Wizard     Patrick Smith  Open Weight  Orthodox  1900-01-01   
3     Patrick Smith        Ray Wizard  Open Weight  Orthodox  1963-08-28   
4     David Levicki     Johnny Rhodes  Open Weight  Orthodox  1900-01-01   
5     Johnny Rhodes     David Levicki  Open Weight  Orthodox  1900-01-01   
6   Thaddeus Luster     Frank Hamaker  Open Weight  Orthodox  1900-01-01   
7     Frank Hamaker   Thaddeus Luster  Open Weight  Orthodox  1900-01-01   
8  Robert Lucarelli      Orlando Wiet  Open Weight  Orthodox  1900-01-01   
9      Orlando Wiet  Robert Lucarelli  Open Weight  Southpaw  1900-01-01   

   method  ...  body_strikes_landed  body_strikes_attempts leg_strikes_landed  \
0     SUB  ...                    0                      0                  0   
1     SUB  ...                    0                      0                  0   
2     SUB  ...                    0                      0                  1   
3     SUB  ...                    1                      1                  0   
4  KO/TKO  ...                    0                      0                  0   
5  KO/TKO  ...                    1                      1                  1   
6     SUB  ...                    0                      0                  0   
7     SUB  ...                    0                      0                  0   
8  KO/TKO  ...                    0                      1                  1   
9  KO/TKO  ...                    1                      1                  0   

  leg_strikes_attempts distance_strikes_landed distance_strikes_attempts  \
0                    2                       0                         3   
1                    0                       0                         0   
2                    1                       1                         1   
3                    0                       0                         0   
4                    0                       1                         2   
5                    1                       1                         1   
6                    0                       0                         0   
7                    0                       1                         1   
8                    3                       2                         6   
9                    0                       1                    

## referee: missing 64
Since there are only 64 missing referee boxes, they can be filled with "Unknown." 

In [10]:
master_dataframe_b["referee"].fillna('Unknown', inplace=True)
display(master_dataframe_b["referee"].head(10))

/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/3301198554.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  master_dataframe_b["referee"].fillna('Unknown', inplace=True)
/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/3301198554.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b["referee

0    John McCarthy
1    John McCarthy
2    John McCarthy
3    John McCarthy
4    John McCarthy
5    John McCarthy
6    John McCarthy
7    John McCarthy
8    John McCarthy
9    John McCarthy
Name: referee, dtype: object

## reach: missing 1283
It seems like most of the reach values are missing from when the data started to first be collected, back in the 1900s. For this value, we will fill it with 0 for now so we can still look through all of the values, but when we are analysing it we will need to separate the older and the newer data from this dataset. 

In [11]:
display(master_dataframe_b["reach"].head(10))
display(master_dataframe_b["reach"].tail(10))
master_dataframe_b["reach"].fillna(0, inplace=True)

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
5   NaN
6   NaN
7   NaN
8   NaN
9   NaN
Name: reach, dtype: float64

13312    72.0
13313    69.0
13314    71.0
13315    70.0
13316    79.0
13317    79.0
13318    80.0
13319    77.0
13320    72.0
13321    70.0
Name: reach, dtype: float64

/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/574418483.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  master_dataframe_b["reach"].fillna(0, inplace=True)
/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/574418483.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b["reach"].fillna(0, i

# height: missing 14
Since there is only 14 missing pieces from height, and for most of the fighters their dob is missing as well, the 14 rows are dropped. 


In [12]:
print(master_dataframe_b["height"][master_dataframe_b["height"].isna()])
print(master_dataframe_b.iloc[[2,7,38,60,214,215,232,288,304,305,318,3066,3244,3476]])
master_dataframe_b.drop([2,7,38,60,214,215,232,288,304,305,318,3066,3244,3476], axis = 0,inplace = True)

2      NaN
7      NaN
38     NaN
60     NaN
214    NaN
215    NaN
232    NaN
288    NaN
304    NaN
305    NaN
318    NaN
3066   NaN
3244   NaN
3476   NaN
Name: height, dtype: float64
            date                                          fight_url  \
2     1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
7     1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
38    1994-09-09  http://ufcstats.com/fight-details/5167a5aec41f...   
60    1995-04-07  http://ufcstats.com/fight-details/fa152ada833d...   
214   1997-05-30  http://ufcstats.com/fight-details/0473ef44197b...   
215   1997-05-30  http://ufcstats.com/fight-details/0473ef44197b...   
232   1997-07-27  http://ufcstats.com/fight-details/f373e50e2fab...   
288   1998-05-15  http://ufcstats.com/fight-details/ea535b46e0f6...   
304   1998-10-16  http://ufcstats.com/fight-details/35b871e0bdc1...   
305   1998-10-16  http://ufcstats.com/fight-details/35b871e0bdc1...   
318   1999-01-08  http://ufcstats.co

/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/3720853074.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b.drop([2,7,38,60,214,215,232,288,304,305,318,3066,3244,3476], axis = 0,inplace = True)


In [13]:
print(master_dataframe_b.head(10))

          date                                          fight_url  \
0   1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
1   1994-03-11  http://ufcstats.com/fight-details/4acab67848e7...   
3   1994-03-11  http://ufcstats.com/fight-details/4b9ae533ccb3...   
4   1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
5   1994-03-11  http://ufcstats.com/fight-details/ccee020be2e8...   
6   1994-03-11  http://ufcstats.com/fight-details/d917c8c7461b...   
8   1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   
9   1994-03-11  http://ufcstats.com/fight-details/3b020d4914b4...   
10  1994-03-11  http://ufcstats.com/fight-details/17ee4caf0698...   
11  1994-03-11  http://ufcstats.com/fight-details/17ee4caf0698...   

                                            event_url  result  \
0   http://ufcstats.com/event-details/a6a9ab5a824e...       0   
1   http://ufcstats.com/event-details/a6a9ab5a824e...       1   
3   http://ufcstats.com/event-details/a6a9ab5

## age: missing 235
Since age is directly correlated to the dob value, and the dob value was filled in with out of range values, the age of everyone who is NaN will be 0.

In [14]:
master_dataframe_b["age"].fillna(0, inplace=True)
display(master_dataframe_b["age"].head(10))

/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/3878346580.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  master_dataframe_b["age"].fillna(0, inplace=True)
/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/3878346580.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b["age"].fillna(0, inp

0     46.0
1      0.0
3     58.0
4      0.0
5      0.0
6      0.0
8      0.0
9      0.0
10     0.0
11     0.0
Name: age, dtype: float64

### Verifying missing values are handled

In [15]:
display(master_dataframe_b.isna().sum())

date                         0
fight_url                    0
event_url                    0
result                       0
fighter                      0
opponent                     0
division                     0
stance                       0
dob                          0
method                       0
total_comp_time              0
round                        0
time                         0
fighter_url                  0
opponent_url                 0
referee                      0
time_format                  0
reach                        0
height                       0
age                          0
knockdowns                   0
sub_attempts                 0
reversals                    0
control                      0
takedowns_landed             0
takedowns_attempts           0
sig_strikes_landed           0
sig_strikes_attempts         0
total_strikes_landed         0
total_strikes_attempts       0
head_strikes_landed          0
head_strikes_attempts        0
body_str

### Standardize Column Names

In [16]:
master_dataframe_b.columns = master_dataframe_b.columns.str.lower().str.replace(' ', '_')

In [17]:
display(master_dataframe_b.columns)

Index(['date', 'fight_url', 'event_url', 'result', 'fighter', 'opponent',
       'division', 'stance', 'dob', 'method', 'total_comp_time', 'round',
       'time', 'fighter_url', 'opponent_url', 'referee', 'time_format',
       'reach', 'height', 'age', 'knockdowns', 'sub_attempts', 'reversals',
       'control', 'takedowns_landed', 'takedowns_attempts',
       'sig_strikes_landed', 'sig_strikes_attempts', 'total_strikes_landed',
       'total_strikes_attempts', 'head_strikes_landed',
       'head_strikes_attempts', 'body_strikes_landed', 'body_strikes_attempts',
       'leg_strikes_landed', 'leg_strikes_attempts', 'distance_strikes_landed',
       'distance_strikes_attempts', 'clinch_strikes_landed',
       'clinch_strikes_attempts', 'ground_strikes_landed',
       'ground_strikes_attempts'],
      dtype='object')

### Converting Data Type
date and dob in Master Dataframe = datetime64[ns]


In [18]:
master_dataframe_b['date'] = pd.to_datetime(master_dataframe_b['date'], errors='coerce')
master_dataframe_b['dob'] = pd.to_datetime(master_dataframe_b['dob'], errors='coerce')

master_dataframe_dtypes = master_dataframe_b.dtypes

master_dataframe_dtypes


/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/493859777.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b['date'] = pd.to_datetime(master_dataframe_b['date'], errors='coerce')
/var/folders/cx/f_jzjtxj4hb2rmkcqgy_py4w0000gn/T/ipykernel_17521/493859777.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  master_dataframe_b['dob'] = pd.to_datetime(master_dataframe_b['dob'], errors='coerce')


date                         datetime64[ns]
fight_url                            object
event_url                            object
result                                int64
fighter                              object
opponent                             object
division                             object
stance                               object
dob                          datetime64[ns]
method                               object
total_comp_time                       int64
round                                 int64
time                                 object
fighter_url                          object
opponent_url                         object
referee                              object
time_format                          object
reach                               float64
height                              float64
age                                 float64
knockdowns                            int64
sub_attempts                          int64
reversals                       

### Inconsistent Data Entry
"orthadox" to "orthodox"

open stance = open_stance

women's weight class = w- 

In [19]:
master_dataframe['stance'] = master_dataframe['stance'].str.lower().replace(
    {'orthadox': 'orthodox', 'open stance': 'open_stance', 'sideways': 'sideways'})

master_stance_unique_cleaned = master_dataframe['stance'].unique()

master_stance_unique_cleaned


array([nan, 'orthodox', 'southpaw', 'open_stance', 'sideways', 'switch'],
      dtype=object)

### Filtering irrelevant colums 

Have not done yet, need to discuss what we do wanna keep and remove

But example:

In [20]:

master_dataframe_cleaned = master_dataframe_b.drop(columns=['event_url', 'fight_url','fighter_url','opponent_url'])
print(master_dataframe_cleaned.columns)



Index(['date', 'result', 'fighter', 'opponent', 'division', 'stance', 'dob',
       'method', 'total_comp_time', 'round', 'time', 'referee', 'time_format',
       'reach', 'height', 'age', 'knockdowns', 'sub_attempts', 'reversals',
       'control', 'takedowns_landed', 'takedowns_attempts',
       'sig_strikes_landed', 'sig_strikes_attempts', 'total_strikes_landed',
       'total_strikes_attempts', 'head_strikes_landed',
       'head_strikes_attempts', 'body_strikes_landed', 'body_strikes_attempts',
       'leg_strikes_landed', 'leg_strikes_attempts', 'distance_strikes_landed',
       'distance_strikes_attempts', 'clinch_strikes_landed',
       'clinch_strikes_attempts', 'ground_strikes_landed',
       'ground_strikes_attempts'],
      dtype='object')


### Saving Cleaned Data

In [21]:
master_dataframe_cleaned.to_csv('cleaned_master_data.csv', index=False)
